File: backend/application/use_cases/authentication/login_use_case.py
from typing import Dict
from datetime import datetime
from domain.entities.user import User
from domain.entities.session import Session
from domain.interfaces.repositories.user_repository import IUserRepository
from domain.interfaces.repositories.session_repository import ISessionRepository
from domain.interfaces.services.token_service import ITokenService
from domain.exceptions.authentication_exceptions import (
    InvalidCredentialsException,
    InactiveUserException
)


class LoginUseCase:
    """Caso de uso para login de usuarios"""
    
    def __init__(
        self,
        user_repository: IUserRepository,
        session_repository: ISessionRepository,
        token_service: ITokenService
    ):
        self.user_repository = user_repository
        self.session_repository = session_repository
        self.token_service = token_service
    
    def execute(self, email: str, password: str) -> Dict:
        """
        Ejecuta el caso de uso de login
        
        Args:
            email: Email del usuario
            password: Contraseña del usuario
        
        Returns:
            Dict con success, token y user
        
        Raises:
            InvalidCredentialsException: Si las credenciales son inválidas
            InactiveUserException: Si el usuario está inactivo
        """
        # 1. Autenticar usuario
        user = self.user_repository.authenticate(email, password)
        if not user:
            raise InvalidCredentialsException("Invalid email or password")
        
        # 2. Verificar que el usuario esté activo
        if not user.is_active:
            raise InactiveUserException("User account is not active")
        
        # 3. Buscar o crear sesión
        session = self.session_repository.find_by_user(user)
        
        if session and session.is_valid() and not session.is_expired():
            # Reutilizar sesión existente
            token = session.token
        else:
            # Crear nueva sesión
            token = self.token_service.generate_token(user)
            session = self.session_repository.create(user, token)
        
        # 4. Retornar resultado
        return {
            "success": True,
            "token": token,
            "user": {
                "_id": user.id,
                "username": user.username,
                "email": user.email
            }
        }

File: backend-dotnet/src/AcademiaDigital.Application/UseCases/Authentication/LoginUseCase.cs
using AcademiaDigital.Domain.Enums;
using AcademiaDigital.Domain.Exceptions;
using AcademiaDigital.Domain.Interfaces.Repositories;
using AcademiaDigital.Domain.Interfaces.Services;

namespace AcademiaDigital.Application.UseCases.Authentication;

public class LoginUseCase(
    IUserRepository userRepository,
    ISessionRepository sessionRepository,
    ITokenService tokenService)
{
    private const int MaxFailedAttempts = 5;
    private static readonly TimeSpan LockoutDuration = TimeSpan.FromMinutes(15);

    public async Task<LoginResult> ExecuteAsync(string email, string password, CancellationToken ct = default)
    {
        var normalizedEmail = email.Trim();
        var existingUser = await userRepository.FindByEmailAsync(normalizedEmail, ct);

        if (existingUser?.LockedUntil is DateTime lockedUntil)
        {
            if (lockedUntil > DateTime.UtcNow)
                throw new AccountLockedException(lockedUntil);

            await userRepository.ResetLoginFailuresAsync(existingUser.Id, ct);
        }

        var user = await userRepository.AuthenticateAsync(normalizedEmail, password, ct);
        if (user is null)
        {
            if (existingUser is not null)
            {
                var updated = await userRepository.RecordFailedLoginAsync(existingUser.Id, MaxFailedAttempts, LockoutDuration, ct);
                if (updated.LockedUntil is DateTime newLockedUntil && updated.FailedLoginAttempts >= MaxFailedAttempts)
                    throw new AccountLockedException(newLockedUntil);
            }

            throw new InvalidCredentialsException();
        }

        if (!user.IsActive)
            throw new InactiveUserException();

        if (user.FailedLoginAttempts > 0 || user.LockedUntil is not null)
            await userRepository.ResetLoginFailuresAsync(user.Id, ct);

        var session = await sessionRepository.FindByUserAsync(user.Id, ct);
        string token;

        if (session != null && !session.IsExpired())
        {
            token = session.Token;
        }
        else
        {
            if (session != null)
                await sessionRepository.DeleteAsync(session, ct);

            token = tokenService.GenerateToken(user);
            await sessionRepository.CreateAsync(user.Id, token, ct);
        }

        return new LoginResult(
            Success: true,
            Token: token,
            User: new UserDto(user.Id, user.Username, user.Email, user.Role));
    }
}

public record LoginResult(bool Success, string Token, UserDto User);
public record UserDto(long Id, string Username, string Email, UserRole Role);
